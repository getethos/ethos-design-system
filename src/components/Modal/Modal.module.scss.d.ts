// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'ModalContent': string;
  'ModalWrapper': string;
  'ModalWrapperActive': string;
}
export const cssExports: CssExports;
export default cssExports;
