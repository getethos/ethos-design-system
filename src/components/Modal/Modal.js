import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import { Portal } from '../Portal'
import styles from './Modal.module.scss'
import useOutsideClick from '../../hooks/a11y/useOutsideClick'
import useBodyScrollLock from '../../hooks/a11y/useBodyScrollLock'
import useTrapFocus from '../../hooks/a11y/useTrapFocus'
import useHideAriaSiblings from '../../hooks/a11y/useHideAriaSiblings'

/**
 * Component renders the internal content of a modal
 *
 * @param {object} props - the component's props
 * @param {React.ReactChildren} props.children - the component's children
 * @param {(val: boolean) => any} props.onDismiss - callback function that onDismisss the
 * show modal state
 *
 * @return {JSX.Element}
 */
const ModalContent = ({
  ariaDescribedBy,
  ariaLabelledBy,
  children,
  onDismiss,
  isOpen,
}) => {
  const modalRef = useRef(null)

  // Hooks! All the thingggs!!!!!
  useOutsideClick(modalRef, () => onDismiss(false))
  useTrapFocus(modalRef, isOpen)
  useBodyScrollLock(isOpen)
  useHideAriaSiblings(modalRef, isOpen)

  return (
    <div
      role="dialog"
      aria-modal="true"
      aria-describedby={ariaDescribedBy && ariaDescribedBy}
      aria-labelledby={ariaLabelledBy && ariaLabelledBy}
      className={styles.ModalContent}
      ref={modalRef}
    >
      {children}
    </div>
  )
}

ModalContent.propTypes = {
  ariaDescribedBy: PropTypes.string,
  ariaLabelledBy: PropTypes.string,
  children: PropTypes.node,
  onDismiss: PropTypes.func,
  isOpen: PropTypes.bool,
}

/**
 * Component renders a modal and it's wrapper
 *
 * @param {object} props - the components props
 * @param {React.ReactChildren} props.children - the component's children
 * @param {boolean} props.isOpen - flag sets the state of the modal
 * @param {(val: boolean) => void} props.onDismiss - handler that onDismisss the state of the modal
 *
 * @return {JSX.Element}
 */
export const Modal = ({
  children,
  onDismiss,
  ariaLabelledBy,
  ariaDescribedBy,
  isOpen = false,
}) => {
  const classes = isOpen ? styles.ModalWrapperActive : styles.ModalWrapper

  /**
   * Handler will set the the modal onDismiss to `false` when the escape key is
   * pressed
   *
   * @private
   *
   * @param {KeyboardEvent} e - the keyboard event
   *
   * @return {void}
   */
  const handleKeyDown = (e) => {
    if (['Escape', 'esc'].includes(e.key) || e.keyCode === 27) {
      onDismiss(false)
    }
  }

  return (
    <Portal id="modal-root">
      {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
      <div
        className={classes}
        onKeyDown={handleKeyDown}
        aria-hidden={!isOpen}
        data-testid="base-modal-container"
      >
        {isOpen && (
          <ModalContent
            onDismiss={onDismiss}
            ariaDescribedBy={ariaDescribedBy}
            ariaLabelledBy={ariaLabelledBy}
            isOpen={isOpen}
          >
            {children}
          </ModalContent>
        )}
      </div>
    </Portal>
  )
}

Modal.propTypes = {
  /** Id of an element that labels the modal */
  ariaLabelledBy: PropTypes.string,
  /** Id of an element that describes the modal */
  ariaDescribedBy: PropTypes.string,
  /** The Modal's children */
  children: PropTypes.node.isRequired,
  /** Boolean that sets the state of the modal */
  isOpen: PropTypes.bool,
  /** handler that onDismisss the state of the modal */
  onDismiss: PropTypes.func.isRequired,
}
