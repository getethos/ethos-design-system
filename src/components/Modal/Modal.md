## Example

```jsx
  import React, { useState, useRef } from 'react'

  function ModalExample() {
    const [isOpen, setIsOpen] = useState(false)
    const onActiveHandler = () => setIsOpen(isOpen);

    return (
      <div>
        <button onClick={() => setIsOpen(true)}>Open Modal</button>
        <Modal showModal={isOpen} toggle={setIsOpen}>
          <div style={{ background: '#fff', padding: '16px 24px', overflow: 'scroll' }}>
            <h1>I am a modal!</h1>
            <p>Here's some text blah balh balh balh</p>
            <button tabIndex="-1" onClick={() => setIsOpen(false)}>Okay</button>
            <button onClick={() => setIsOpen(false)}>Cancel</button>
          </div>
        </Modal>
      </div>
    )
  }

  <ModalExample />
```