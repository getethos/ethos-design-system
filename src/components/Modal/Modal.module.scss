@import '../Variables.scss';

.ModalWrapper {
  display: none;

  &Active {
    align-items: center;
    background: var(--GraySecondary--translucent);
    display: flex;
    height: 100vh;
    justify-content: center;
    left: 0;
    position: fixed;
    top: 0;
    width: 100vw;
    z-index: $z-index-modal;
  }
}

.ModalContent {
  position: relative;
  overflow: auto;
}
