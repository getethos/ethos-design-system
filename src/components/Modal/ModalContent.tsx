import * as React from 'react'
import { useRef } from 'react'
import useBodyScrollLock from '../../hooks/a11y/useBodyScrollLock'
import useHideAriaSiblings from '../../hooks/a11y/useHideAriaSiblings'
import useOutsideClick from '../../hooks/a11y/useOutsideClick'
import useTrapFocus from '../../hooks/a11y/useTrapFocus'
import styles from './Modal.module.scss'
type ModalContentProps = {
  ariaDescribedBy?: string
  ariaLabelledBy?: string
  onDismiss?: (...args: any[]) => any
  isOpen?: boolean
}
/**
 * Component renders the internal content of a modal
 *
 * @param {object} props - the component's props
 * @param {React.ReactChildren} props.children - the component's children
 * @param {(val: boolean) => any} props.onDismiss - callback function that onDismisss the
 * show modal state
 *
 * @return {JSX.Element}
 */
export const ModalContent: React.FC<ModalContentProps> = ({
  ariaDescribedBy,
  ariaLabelledBy,
  children,
  onDismiss = () => {},
  isOpen = false,
}) => {
  const modalRef = useRef(null)
  // Hooks! All the thingggs!!!!!
  useOutsideClick(modalRef, () => onDismiss(false))
  useTrapFocus(modalRef, isOpen)
  useBodyScrollLock(isOpen)
  useHideAriaSiblings(modalRef, isOpen)
  return (
    <div
      role="dialog"
      aria-modal="true"
      aria-describedby={ariaDescribedBy && ariaDescribedBy}
      aria-labelledby={ariaLabelledBy && ariaLabelledBy}
      className={styles.ModalContent}
      ref={modalRef}
    >
      {children}
    </div>
  )
}
