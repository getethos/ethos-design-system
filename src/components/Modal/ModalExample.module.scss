.SmallModal {
  background: #fff;
  padding: 16px 24px;
  width: 400px;
}

.TakeoverModal {
  background: #fff;
  padding: 16px 24px;
  overflow: auto;
  width: 100vw;
  height: 100vh;
}
