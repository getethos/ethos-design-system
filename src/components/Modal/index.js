export { Modal } from './Modal.js'
