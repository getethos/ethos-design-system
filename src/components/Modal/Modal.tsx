import * as React from 'react'
import { Portal } from '../Portal'
import styles from './Modal.module.scss'
import { ModalContent } from './ModalContent'
type ModalProps = {
  ariaLabelledBy?: string
  ariaDescribedBy?: string
  isOpen?: boolean
  onDismiss: (...args: any[]) => any
}
/**
 * Component renders a modal and it's wrapper
 *
 * @param {object} props - the components props
 * @param {React.ReactChildren} props.children - the component's children
 * @param {boolean} props.isOpen - flag sets the state of the modal
 * @param {(val: boolean) => void} props.onDismiss - handler that onDismisss the state of the modal
 *
 * @return {JSX.Element}
 */
export const Modal: React.SFC<ModalProps> = ({
  children,
  onDismiss,
  ariaLabelledBy,
  ariaDescribedBy,
  isOpen = false,
}) => {
  const classes = isOpen ? styles.ModalWrapperActive : styles.ModalWrapper
  /**
   * Handler will set the the modal onDismiss to `false` when the escape key is
   * pressed
   *
   * @private
   *
   * @param {KeyboardEvent} e - the keyboard event
   *
   * @return {void}
   */
  const handleKeyDown = (e) => {
    if (['Escape', 'esc'].includes(e.key) || e.keyCode === 27) {
      onDismiss(false)
    }
  }
  return (
    <Portal id="modal-root">
      <div
        className={classes}
        onKeyDown={handleKeyDown}
        aria-hidden={!isOpen}
        data-testid="base-modal-container"
      >
        {isOpen && (
          <ModalContent
            onDismiss={onDismiss}
            ariaDescribedBy={ariaDescribedBy}
            ariaLabelledBy={ariaLabelledBy}
            isOpen={isOpen}
          >
            {children}
          </ModalContent>
        )}
      </div>
    </Portal>
  )
}
