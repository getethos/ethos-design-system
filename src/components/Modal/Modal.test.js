import React from 'react'
import ReactDOM from 'react-dom'
import { Modal } from './Modal'
import { render, fireEvent, act } from '@testing-library/react'

describe('Modal', () => {
  beforeAll(() => {
    // createPortal needs to be mocked
    // SEE: https://github.com/facebook/react/issues/11565
    ReactDOM.createPortal = jest.fn((element) => {
      return element
    })
  })

  afterEach(() => {
    ReactDOM.createPortal.mockClear()
  })

  it("Render's a modal", () => {
    const isOpen = true
    const onDismissMock = jest.fn()
    window.scrollTo = jest.fn()
    const ariaDescribedBy = 'some-id'
    const ariaLabelledBy = 'some-other-id'

    const props = {
      ariaDescribedBy,
      ariaLabelledBy,
      isOpen,
      onDismiss: onDismissMock,
    }

    const tree = render(
      <Modal {...props}>
        <h1 id={ariaLabelledBy}>Ima modal</h1>
        <p id={ariaDescribedBy}>some more content</p>
      </Modal>
    )

    expect(tree).toMatchSnapshot()
  })

  it("Fire's `onDismiss` when the `escape` key is pressed", () => {
    const isOpen = true
    const onDismissMock = jest.fn()
    window.scrollTo = jest.fn()

    const props = {
      isOpen,
      onDismiss: onDismissMock,
    }

    const { getByTestId } = render(
      <Modal {...props}>
        <h1>Ima modal</h1>
      </Modal>
    )

    const modal = getByTestId('base-modal-container')

    act(() => {
      fireEvent.keyDown(modal, { key: 'Escape', keyCode: 27 })
    })

    expect(onDismissMock).toHaveBeenCalled()
  })
})
