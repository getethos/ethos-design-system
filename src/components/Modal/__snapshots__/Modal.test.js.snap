// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Render's a modal 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    style="overflow: hidden; position: fixed; width: 100%; top: 0px;"
  >
    <div>
      <div
        class="ModalBackdrop"
      />
      <div
        aria-hidden="false"
        class="ModalWrapperActive"
        data-testid="base-modal-container"
      >
        <div
          aria-describedby="some-id"
          aria-labelledby="some-other-id"
          aria-modal="true"
          class="ModalContent"
          role="dialog"
        >
          <h1
            id="some-other-id"
          >
            Ima modal
          </h1>
          <p
            id="some-id"
          >
            some more content
          </p>
        </div>
      </div>
    </div>
    <div
      id="modal-root"
    />
  </body>,
  "container": <div>
    <div
      class="ModalBackdrop"
    />
    <div
      aria-hidden="false"
      class="ModalWrapperActive"
      data-testid="base-modal-container"
    >
      <div
        aria-describedby="some-id"
        aria-labelledby="some-other-id"
        aria-modal="true"
        class="ModalContent"
        role="dialog"
      >
        <h1
          id="some-other-id"
        >
          Ima modal
        </h1>
        <p
          id="some-id"
        >
          some more content
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
